compiler/zrc_preprocessor/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+    clippy::multiple_inherent_impl,
+    clippy::multiple_unsafe_ops_per_block,
+    clippy::non_ascii_literal,
+    clippy::redundant_type_annotations,
+    clippy::rest_pat_in_fully_bound_structs,
+    clippy::same_name_method,
+    clippy::semicolon_inside_block,
+    clippy::unseparated_literal_suffix,
+    clippy::implicit_clone,
+    clippy::todo,
+    clippy::undocumented_unsafe_blocks,
+    clippy::unimplemented,
+    clippy::unneeded_field_pattern,
+    clippy::wildcard_enum_match_arm,
+    let_underscore_drop,
+    macro_use_extern_crate,
+    missing_debug_implementations,
+    non_exhaustive_omitted_patterns,
+    unsafe_op_in_unsafe_fn,
+    unused_crate_dependencies,
+    variant_size_differences,
+    unused_qualifications,
+    clippy::unwrap_used,
+    clippy::print_stderr,
+    clippy::print_stdout
+)]
+#![allow(
+    clippy::multiple_crate_versions,
+    clippy::cargo_common_metadata,
+    clippy::module_name_repetitions,
+    clippy::doc_comment_double_space_linebreaks
+)]
+
+use std::{
+    collections::HashSet,
+    fs,
+    path::{Path, PathBuf},
+};
+
+use zrc_diagnostics::{Diagnostic, DiagnosticKind};
+use zrc_utils::span::Span;
+
+/// Represents a chunk of source code with its metadata
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SourceChunk {
+    /// The file name where this chunk originates from
+    pub file_name: String,
+    /// The starting line number (1-indexed) in the original file
+    pub start_line: usize,
+    /// The content of this chunk
+    pub content: String,
+}
+
+impl SourceChunk {
+    /// Create a new source chunk
+    #[must_use]
+    pub const fn new(file_name: String, start_line: usize, content: String) -> Self {
+        Self {
+            file_name,
+            start_line,
+            content,
+        }
+    }
+}
+
+/// Process a Zirco source file with preprocessing directives
+///
+/// # Arguments
+/// * `base_path` - The directory to resolve relative includes from
+/// * `file_name` - The name of the file being processed
+/// * `content` - The content of the file to preprocess
+///
+/// # Errors
+/// Returns an error if:
+/// - An included file cannot be found
+/// - An included file cannot be read
+/// - A preprocessing directive is malformed
+#[allow(clippy::result_large_err)]
+pub fn preprocess(
+    base_path: &Path,
+    file_name: &str,
+    content: &str,
+) -> Result<Vec<SourceChunk>, Diagnostic> {
+    let mut chunks = Vec::new();
+    let mut pragma_once_files = HashSet::new();
+
+    preprocess_internal(
+        base_path,
+        file_name,
+        content,
+        &mut chunks,
+        &mut pragma_once_files,
+    )?;
+
+    Ok(chunks)
+}
+
+/// Internal recursive preprocessing function
+#[allow(clippy::too_many_lines)]
+#[allow(clippy::result_large_err)]
+fn preprocess_internal(
+    base_path: &Path,
+    file_name: &str,
+    content: &str,
+    chunks: &mut Vec<SourceChunk>,
+    pragma_once_files: &mut HashSet<PathBuf>,
```

Comment:
Generate and pass a `PreprocessorCtx` instead of these separate files. This is idiomatic zrc design per cg and tck.